From a916232763016e6c0d68f9073822b8bee1094b72 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lo=C3=AFc=20Le=20Page?= <llepage@igalia.com>
Date: Tue, 16 Jun 2026 16:34:31 +0200
Subject: [PATCH] souphttpsrc: fix race conditions when looping

There are 2 race conditions happenning when seeking-flush to 0 after
receiving the EOS with `souphttpsrc`:
- When the request is cancelled at the moment of the flush, the
  `SoupMessage` is not reset, which may lead to not correctly start the
  next request and so we may be waiting forever on the `src->session_cond`.
  This patch systematically clears the dropped `SoupMessage` when the
  request is cancelled and the element is flushing.
- When the seek-flush happens, the `src->cancellable` may be reset while a
  request is still processing. As the `g_cancellable_reset()` documentation
  states: "If cancellable is currently in use by any cancellable operation
  then the behavior of this function is undefined". To avoid this
  situation, this patch uses specific cancellables for each individual
  Send/Read request. The top-level `src->cancellable` is only used to manage
  the cancellation state and to trigger the request specific
  dropped-after-use cancellable.
---
 .../ext/soup/gstsouphttpsrc.c                 | 28 +++++++++++++++++--
 1 file changed, 25 insertions(+), 3 deletions(-)

diff --git a/subprojects/gst-plugins-good/ext/soup/gstsouphttpsrc.c b/subprojects/gst-plugins-good/ext/soup/gstsouphttpsrc.c
index 0c1db5bea3..422f9a3287 100644
--- a/subprojects/gst-plugins-good/ext/soup/gstsouphttpsrc.c
+++ b/subprojects/gst-plugins-good/ext/soup/gstsouphttpsrc.c
@@ -1960,6 +1960,7 @@ struct GstSoupSendSrc
 {
   GstSoupHTTPSrc *src;
   GError *error;
+  GCancellable *soup_cancellable;
 };
 
 static void
@@ -1976,6 +1977,7 @@ _session_send_cb (GObject * source, GAsyncResult * res, gpointer user_data)
 
   if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
     src->headers_ret = GST_FLOW_FLUSHING;
+    g_clear_object (&src->msg);
   } else {
     src->headers_ret = gst_soup_http_src_got_headers (src, src->msg);
   }
@@ -1995,8 +1997,8 @@ _session_send_idle_cb (gpointer user_data)
   struct GstSoupSendSrc *msrc = user_data;
   GstSoupHTTPSrc *src = msrc->src;
 
-  _soup_session_send_async (src->session->session, src->msg, src->cancellable,
-      _session_send_cb, msrc);
+  _soup_session_send_async (src->session->session, src->msg,
+      msrc->soup_cancellable, _session_send_cb, msrc);
 
   return FALSE;
 }
@@ -2008,12 +2010,16 @@ gst_soup_http_src_send_message (GstSoupHTTPSrc * src)
   GstFlowReturn ret;
   GSource *source;
   struct GstSoupSendSrc msrc;
+  gulong cancel_id;
 
   g_return_val_if_fail (src->msg != NULL, GST_FLOW_ERROR);
   g_assert (src->input_stream == NULL);
 
   msrc.src = src;
   msrc.error = NULL;
+  msrc.soup_cancellable = g_cancellable_new ();
+  cancel_id = g_cancellable_connect (src->cancellable,
+      G_CALLBACK (g_cancellable_cancel), msrc.soup_cancellable, NULL);
 
   source = g_idle_source_new ();
 
@@ -2026,6 +2032,9 @@ gst_soup_http_src_send_message (GstSoupHTTPSrc * src)
   while (!src->input_stream && !msrc.error)
     g_cond_wait (&src->session_cond, &src->session_mutex);
 
+  g_cancellable_disconnect (src->cancellable, cancel_id);
+  g_object_unref (msrc.soup_cancellable);
+
   ret = src->headers_ret;
 
   if (ret != GST_FLOW_OK) {
@@ -2117,6 +2126,7 @@ gst_soup_http_src_do_request (GstSoupHTTPSrc * src, const gchar * method)
 
   if (g_cancellable_is_cancelled (src->cancellable)) {
     GST_INFO_OBJECT (src, "interrupted");
+    g_clear_object (&src->msg);
     return GST_FLOW_FLUSHING;
   }
 
@@ -2218,6 +2228,7 @@ struct GstSoupReadResult
   void *buffer;
   gsize bufsize;
   gssize nbytes;
+  GCancellable *soup_cancellable;
 };
 
 static void
@@ -2240,7 +2251,7 @@ _session_read_idle_cb (gpointer user_data)
   struct GstSoupReadResult *res = user_data;
 
   g_input_stream_read_async (res->src->input_stream, res->buffer,
-      res->bufsize, G_PRIORITY_DEFAULT, res->src->cancellable,
+      res->bufsize, G_PRIORITY_DEFAULT, res->soup_cancellable,
       _session_read_cb, res);
 
   return FALSE;
@@ -2254,6 +2265,7 @@ gst_soup_http_src_read_buffer (GstSoupHTTPSrc * src, GstBuffer ** outbuf)
   GstBaseSrc *bsrc;
   GstFlowReturn ret;
   GSource *source;
+  gulong cancel_id;
 
   bsrc = GST_BASE_SRC_CAST (src);
 
@@ -2273,6 +2285,9 @@ gst_soup_http_src_read_buffer (GstSoupHTTPSrc * src, GstBuffer ** outbuf)
   res.bufsize = mapinfo.size;
   res.error = NULL;
   res.nbytes = -1;
+  res.soup_cancellable = g_cancellable_new ();
+  cancel_id = g_cancellable_connect (src->cancellable,
+      G_CALLBACK (g_cancellable_cancel), res.soup_cancellable, NULL);
 
   source = g_idle_source_new ();
 
@@ -2296,12 +2311,15 @@ gst_soup_http_src_read_buffer (GstSoupHTTPSrc * src, GstBuffer ** outbuf)
     GstFlowReturn ret = GST_FLOW_CUSTOM_ERROR;
     if (g_error_matches (res.error, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
       ret = GST_FLOW_FLUSHING;
+      g_clear_object (&src->msg);
     } else {
       GST_ERROR_OBJECT (src, "Got error from libsoup: %s", res.error->message);
     }
     g_error_free (res.error);
     gst_buffer_unmap (*outbuf, &mapinfo);
     gst_buffer_unref (*outbuf);
+    g_cancellable_disconnect (src->cancellable, cancel_id);
+    g_object_unref (res.soup_cancellable);
     return ret;
   }
 
@@ -2372,6 +2390,9 @@ gst_soup_http_src_read_buffer (GstSoupHTTPSrc * src, GstBuffer ** outbuf)
 
   g_clear_error (&res.error);
 
+  g_cancellable_disconnect (src->cancellable, cancel_id);
+  g_object_unref (res.soup_cancellable);
+
   return ret;
 }
 
@@ -2383,6 +2404,7 @@ _session_stream_clear_cb (gpointer user_data)
   g_mutex_lock (&src->session_mutex);
 
   g_clear_object (&src->input_stream);
+  g_clear_object (&src->msg);
 
   g_cond_signal (&src->session_cond);
   g_mutex_unlock (&src->session_mutex);
-- 
2.43.0

